mount/losetup_linux.go

```diff
@@ -0,0 +1,229 @@
+*/
+
+package mount
+
+import (
+	"fmt"
+	"math/rand"
+	"os"
+	"strings"
+	"syscall"
+	"time"
+	"unsafe"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	loopControlPath = "/dev/loop-control"
+	loopDevFormat   = "/dev/loop%d"
+
+	// According to util-linux/include/loopdev.h
+	ioctlSetFd       = 0x4C00
+	ioctlClrFd       = 0x4C01
+	ioctlSetStatus64 = 0x4C04
+	ioctlGetFree     = 0x4C82
+
+	loFlagsReadonly = 1
+	//loFlagsUseAops   = 2
+	loFlagsAutoclear = 4
+	//loFlagsPartScan  = 8
+	loFlagsDirectIO = 16
+
+	ebusyString = "device or resource busy"
+)
+
+// parameters to control loop device setup
+type LoopParams struct {
+	// Loop device should forbid write
+	Readonly bool
+	// Loop device is automatically cleared by kernel when the
+	// last opener closes it
+	Autoclear bool
+	// Use direct IO to access the loop backing file
+	Direct bool
+}
+
+// struct loop_info64 in util-linux/include/loopdev.h
+type loopInfo struct {
+	/*
+		device         uint64
+		inode          uint64
+		rdevice        uint64
+		offset         uint64
+		sizelimit      uint64
+		number         uint32
+		encryptType    uint32
+		encryptKeySize uint32
+	*/
+	_        [13]uint32
+	flags    uint32
+	fileName [64]byte
+	/*
+		cryptName  [64]byte
+		encryptKey [32]byte
+		init       [2]uint64
+	*/
+	_ [112]byte
+}
+
+func ioctl(fd, req, args uintptr) (uintptr, uintptr, error) {
+	r1, r2, errno := syscall.Syscall(syscall.SYS_IOCTL, fd, req, args)
+	if errno != 0 {
+		return 0, 0, errno
+	}
+
+	return r1, r2, nil
+}
+
+func getFreeLoopDev() (uint32, error) {
+	ctrl, err := os.OpenFile(loopControlPath, os.O_RDWR, 0)
+	if err != nil {
+		return 0, errors.Errorf("could not open %v: %v", loopControlPath, err)
+	}
+	defer ctrl.Close()
+	num, _, err := ioctl(ctrl.Fd(), ioctlGetFree, 0)
+	if err != nil {
+		return 0, errors.Wrap(err, "could not get free loop device")
+	}
+	return uint32(num), nil
+}
+
+func setupLoopDev(backingFile, loopDev string, param LoopParams) (devFile *os.File, err error) {
+	// 1. Open backing file and loop device
+	oflags := os.O_RDWR
+	if param.Readonly {
+		oflags = os.O_RDONLY
+	}
+	back, err := os.OpenFile(backingFile, oflags, 0)
+	if err != nil {
+		return nil, errors.Errorf("could not open backing file: %v", err)
+	}
+	defer back.Close()
+
+	loopFile, err := os.OpenFile(loopDev, oflags, 0)
+	if err != nil {
+		return nil, errors.Errorf("could not open loop device: %v", err)
+	}
+	defer func() {
+		if err != nil {
+			loopFile.Close()
+		}
+	}()
+
+	// 2. Set FD
+	if _, _, err = ioctl(loopFile.Fd(), ioctlSetFd, back.Fd()); err != nil {
+		return nil, errors.Errorf("could not set loop fd: %v", err)
+	}
+
+	// 3. Set Info
+	info := loopInfo{}
+	copy(info.fileName[:], []byte(backingFile))
+	if param.Readonly {
+		info.flags |= loFlagsReadonly
+	}
+	if param.Autoclear {
+		info.flags |= loFlagsAutoclear
+	}
+	if param.Direct {
+		info.flags |= loFlagsAutoclear
+	}
+	if _, _, err := ioctl(loopFile.Fd(), ioctlSetStatus64, uintptr(unsafe.Pointer(&info))); err != nil {
```

Comment:
What's the benefit of having this losetup code vs the old package ?

Reply:
The main purpose of the PR is to enable `mount` package to handle loopback mounts so that I can use it for the rawblock snapshotter. Maintainers also want to consolidate the two losetup implementation so I added the same APIs to replace the `losetup` command line wrappers. The main difference between them is calling the `losetup` binary or not.

Reply:
Sorry, I meant why not just reuse existing package. Potentially it may be a lot harder to maintain and support, since Go is not the best fit for low level programming.

Reply:
Because it is not possible to create a autoclear loopback device with `losetup` command line. The workflow I want is:
1. open the loop device
2. set its backend and set the `autoclear` flag
3. mount the loopback device with proper file system
4. close the initial open file descriptor that we got in step 1

This is basically what you get with `mount -o loop` but it is not possible with the `losetup` command line. And the benefit is worthwhile -- we don't have to care about deleting the loopback device after unmounting the file system, nor if anything goes wrong in the middle before we mount the file system. The kernel takes care of cleaning up the loopback device once the last opener to it closes its open file descriptor.

Reply:
And we're already calling mount syscall instead of the `mount` binary in the mount package. Doing the same for loopback device follows the same philosophy.

Reply:
> Because it is not possible to create an autoclear loopback device with losetup command line.

hm. As far as I understand `losetup` does that with `--detach` flag.

```
-d, --detach loopdev...
              Detach the file or device associated with the specified loop
              device(s). Note that since Linux v3.7 kernel uses "lazy device
              destruction".  The detach operation does not return EBUSY
              error anymore if device is actively used by system, but it is
              marked by autoclear flag and destroyed later.
```

http://man7.org/linux/man-pages/man8/losetup.8.html

Reply:
It is the detach action, not setting a loop device as autoclear. With autoclear, the loop device is detached automatically **after** it is unmounted -- that means we do not need to call `losetup -d` after unmounting a loop device.

If we use `losetup` command line, the workflow is like:
1. `losetup -f <backing-file>`
2. mount <loop-device> with a file system
3. unmount the file system
4. `losetup -d <loop-device>`

Whereas with autoclear, we do:
1. `mount -o loop <backing-file>` alike with `mount` package
2. umount the file system

Reply:
Or do you mean that we should do the following?
1. `losetup -f <backing-file>`
2. mount with a file system
3. `losetup -d <loop-device>`
4. unmount the file system

That still doesn't look as neat as just doing
1. `mount -o loop <backing-file>` alike with mount package
2. umount the file system